src/core/font_substitutions.js

```diff
@@ -0,0 +1,458 @@
+        append: "Italic",
+      },
+      style: ITALIC,
+      fallback: "Helvetica-Italic",
+    },
+  ],
+  [
+    "ArialNarrow-BoldItalic",
+    {
+      local: {
+        alias: "ArialNarrow",
+        append: "Bold Italic",
+      },
+      style: BOLDITALIC,
+      fallback: "Helvetica-BoldItalic",
+    },
+  ],
+  [
+    "Calibri",
+    {
+      prepend: ["Calibri", "Carlito"],
+      style: NORMAL,
+      fallback: "Helvetica",
+    },
+  ],
+  [
+    "Calibri-Bold",
+    {
+      local: {
+        alias: "Calibri",
+        append: "Bold",
+      },
+      style: BOLD,
+      fallback: "Helvetica-Bold",
+    },
+  ],
+  [
+    "Calibri-Italic",
+    {
+      local: {
+        alias: "Calibri",
+        append: "Italic",
+      },
+      style: ITALIC,
+      fallback: "Helvetica-Italic",
+    },
+  ],
+  [
+    "Calibri-BoldItalic",
+    {
+      local: {
+        alias: "Calibri",
+        append: "Bold Italic",
+      },
+      style: BOLDITALIC,
+      fallback: "Helvetica-BoldItalic",
+    },
+  ],
+]);
+
+/**
+ * Create the src path to use to load a font (see FontFace).
+ * @param {Array<String>} prepend A list of font names to search first.
+ * @param {Array<String>|Object} local A list of font names to search. If an
+ *   Object is passed, then local.alias is the name of an other substition font
+ *   and local.append is a String to append to the list of fonts in the alias.
+ *   For example if local.alias is "Foo" and local.append is "Bold" then the
+ *   list of fonts will be "FooSubst1 Bold", "FooSubst2 Bold", etc.
+ * @returns an String with the local fonts.
+ */
+function makeLocal(prepend, local) {
+  let append = "";
+  if (!Array.isArray(local)) {
+    // We are getting our list of fonts in the alias and we'll append Bold,
+    // Italic or both.
+    append = ` ${local.append}`;
+    local = substitutionMap.get(local.alias).local;
+  }
+  let prependedPaths = "";
+  if (prepend) {
+    prependedPaths = prepend.map(name => `local(${name})`).join(",") + ",";
+  }
+  return (
+    prependedPaths + local.map(name => `local(${name}${append})`).join(",")
+  );
+}
+
+/**
+ * Get a font substitution for a given font.
+ * The general idea is to have enough information to create a CSS rule like
+ * this:
+ *   @font-face {
+ *    font-family: 'Times';
+ *    src: local('Times New Roman'), local('Subst1'), local('Subst2'),
+ *         url(.../TimesNewRoman.ttf)
+ *    font-weight: normal;
+ *    font-style: normal;
+ *   }
+ * or use the FontFace API.
+ *
+ * @param {Object} idFactory The ids factory.
+ * @param {String} localFontPath Path to the fonts directory.
+ * @param {String} baseFontName The font name to be substituted.
+ * @param {String} standardFontName The standard font name to use if the base
+ *   font is not available.
+ * @returns an Object with the CSS, the loaded name, the src and the style.
+ */
+function getFontSubstitution(
+  idFactory,
+  localFontPath,
+  baseFontName,
+  standardFontName
+) {
```

Comment:
Given the `ArialBlack,Bold` problem mentioned in https://github.com/mozilla/pdf.js/pull/16363#issuecomment-1540787068, should we perhaps just normalize the name here to avoid having to add loads of alias-entries in the font-substitution map?
```suggestion
) {
  baseFontName = normalizeFontName(baseFontName);
```
Which would be consistent with how we handle names of non-embedded fonts elsewhere see also https://github.com/mozilla/pdf.js/blob/1a2e787e47cfb6178f5eeedc608fd9bd571cf0b3/src/core/fonts_utils.js#L183-L185

Reply:
Yep, I do almost the same but with a small variation, tell me wdyt about ?

I found a cache issue in `font_loader`, consequently I removed `#systemFont` set.
The key was only based on `src` which wasn't a good idea because if two fonts have the same `src` (and even the same style) then g_d0_f2 was not loaded (when g_d0_f1 was) and in canvas::setFont we're using g_d0_f2 which is not defined.

In my work on refactoring the text layer I disabled the font cache clearing because we need to have the font for the text layer.

Reply:
> Yep, I do almost the same but with a small variation, tell me wdyt about ?

Do we actually *need* to diverge here, since that small inconsistency feels like it could potentially cause issues (or at least surprises) later on?

Please note that elsewhere when dealing with non-embedded fonts we're using `normalizeFontName`, see e.g. https://github.com/mozilla/pdf.js/blob/1a2e787e47cfb6178f5eeedc608fd9bd571cf0b3/src/core/standard_fonts.js#L883-L887 and https://github.com/mozilla/pdf.js/blob/1a2e787e47cfb6178f5eeedc608fd9bd571cf0b3/src/core/fonts.js#L1134-L1139

Reply:
I find a bit strange to remove the spaces since it could lead to concatenate some words and then create a new one.
That said I'm fine with using normalizeFontName.

Reply:
If using `normalizeFontName` works I'd really prefer a that, since a *consistent* handling of non-embedded font names should limit surprises/problems later.